tests/decomp/040-bzero/bzero.lst.exp.clean

```diff
@@ -2,10 +2,11 @@
 void bzero()
 {
   if ($a3_0 != 0) {
-    $a6 = $a2_0;
     if (($a3_0 & BIT(0)) != 0) {
       *(u8*)$a2_0 = 0x0;
       $a6 = $a2_0 + 0x1;
+    } else {
+      $a6 = $a2_0;
```

Comment:
So, this is where it really goes not how I'd like. But wait again, even the older code was not how I'd like it to be. Instead, it should be:

~~~
    $a6 = $a2_0;
     if (($a3_0 & BIT(0)) != 0) {
       $a6 += 0x1;
    }
~~~

But again, it's not like that, and the new pass transforms it in that way only because it's not how it ideally should be. If we fix that, this transform won't happen either. In the meantime, it's maybe only a little bit worse.

So, I'm -0.1 on this change.

Note also that this is an interesting example of the need to limit propagation. Common sense says that we should not propagate "complex" expression to more than 1 place, but here we see that even simple expressions should not be propagated, if "places" affected participate in peculiar dataflow relations.

Reply:
> Note also that this is an interesting example of the need to limit propagation. Common sense says that we should not propagate "complex" expression to more than 1 place, but here we see that even simple expressions should not be propagated, if "places" affected participate in peculiar dataflow relations.

This shows that the primitive, threshold-based solution given in Van Emmerik's thesis (page 226) isn't good enough for real-world programs. Ilfak Guilfanov confirmed that his Hexrays decompiler relies on several heuristics that guide expression propagation. We could collect and examine the cases where expression propagation isn't desired and try to infer some common rules.
And the user should get the opportunity to undo/redo expression propagations interactively.

Reply:
> We could collect and examine the cases where expression propagation isn't desired and try to infer some common rules.

Yep, feel free to start such a ticket.

Reply:
Actually, started that myself: https://github.com/pfalcon/ScratchABlock/issues/21